<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Load and combine data for use with other mmgenome2 functions — mmload • mmgenome2</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>


<!-- docsearch -->
<script src="../docsearch.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous" />
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>



<meta property="og:title" content="Load and combine data for use with other mmgenome2 functions — mmload" />
<meta property="og:description" content="Loads, validates and combines multiple aspects of metagenome data into one dataframe for use with all mmgenome2 functions, including scaffold assembly sequences, scaffold coverage, essential genes, taxonomy, and more." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-93913019-3"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93913019-3');
</script>


  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mmgenome2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/mmgenome2.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/kasperskytte/mmgenome2">
    <span class="fas fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
      <form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
        </div>
      </form>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Load and combine data for use with other mmgenome2 functions</h1>
    <small class="dont-index">Source: <a href='https://github.com/kasperskytte/mmgenome2/blob/master/R/mmload.R'><code>R/mmload.R</code></a></small>
    <div class="hidden name"><code>mmload.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Loads, validates and combines multiple aspects of metagenome data into one dataframe for use with all mmgenome2 functions, including scaffold assembly sequences, scaffold coverage, essential genes, taxonomy, and more.</p>
    </div>

    <pre class="usage"><span class='fu'>mmload</span><span class='op'>(</span>
  <span class='va'>assembly</span>,
  coverage <span class='op'>=</span> <span class='cn'>NULL</span>,
  essential_genes <span class='op'>=</span> <span class='cn'>NULL</span>,
  taxonomy <span class='op'>=</span> <span class='cn'>NULL</span>,
  additional <span class='op'>=</span> <span class='cn'>NULL</span>,
  kmer_pca <span class='op'>=</span> <span class='cn'>FALSE</span>,
  kmer_BH_tSNE <span class='op'>=</span> <span class='cn'>FALSE</span>,
  kmer_size <span class='op'>=</span> <span class='fl'>4L</span>,
  verbose <span class='op'>=</span> <span class='cn'>TRUE</span>,
  <span class='va'>...</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>assembly</th>
      <td><p>(<em>required</em>) A character string with the path to the assembly FASTA file, or the assembly as already loaded with <code>readDNAStringSet</code>.</p></td>
    </tr>
    <tr>
      <th>coverage</th>
      <td><p>(<em>required</em>) A path to a folder to scan for coverage files, or otherwise a named <code>vector</code>, <code>data.frame</code>, or a <code>list</code> hereof containing coverage of each scaffold. The prefix <code>"cov_"</code> will be appended to all coverage column names in the output so that <code><a href='mmstats.html'>mmstats</a></code> and <code><a href='mmplot_cov_profiles.html'>mmplot_cov_profiles</a></code> know which columns are coverage columns.</p><dl>
  <dt><code>vector</code></dt><dd><p>If provided as a vector, the elements of the vector must be named by the scaffold names exactly matching those of the assembly.</p></dd>
  <dt><code>data.frame</code></dt><dd><p>If provided as a dataframe, the first column must contain the scaffold names exactly matching those of the assembly, and any additional column(s) contain coverage of each scaffold.</p></dd>
  <dt><code>list</code></dt><dd><p>If provided as a list, it must contain any number of <code>vector</code>'s or <code>data.frame</code>'s as described above. If names are assigned to the objects in the list, then they will be used as column names in the output (does not apply to any dataframes that may have more than 2 columns, however).</p></dd>
  <dt><code>path</code></dt><dd><p>If a path to a folder is provided, then all files with filenames ending with <code>"_cov"</code> will be loaded (by the <code><a href='https://Rdatatable.gitlab.io/data.table/reference/fread.html'>fread</a></code> function) into a list of <code>data.frame</code>'s and treated as if a <code>list</code> of <code>data.frame</code>'s were provided. The filenames (stripped from extension and <code>"_cov"</code>) will then be used as column names in the output. <strong>Note: only the first 2 columns will be used in the loaded files!</strong></p></dd>

</dl></td>
    </tr>
    <tr>
      <th>essential_genes</th>
      <td><p>Either a path to a CSV file (comma-delimited ",") containing the essential genes, or a 2-column dataframe with scaffold names in the first column and gene ID's in the second. Can contain duplicates. (<em>Default: </em> <code>NULL</code>)</p></td>
    </tr>
    <tr>
      <th>taxonomy</th>
      <td><p>A dataframe containing taxonomy assigned to the scaffolds. The first column must contain the scaffold names. (<em>Default: </em> <code>NULL</code>)</p></td>
    </tr>
    <tr>
      <th>additional</th>
      <td><p>A dataframe containing any additional data. The first column must contain the scaffold names. (<em>Default: </em> <code>NULL</code>)</p></td>
    </tr>
    <tr>
      <th>kmer_pca</th>
      <td><p>(<em>Logical</em>) Perform Principal Components Analysis of kmer nucleotide frequencies (kmer size defined by <code>kmer_size</code>) of each scaffold and merge the scores of the 3 most significant axes. (<em>Default: </em> <code>FALSE</code>)</p></td>
    </tr>
    <tr>
      <th>kmer_BH_tSNE</th>
      <td><p>(<em>Logical</em>) Calculate Barnes-Hut t-Distributed Stochastic Neighbor Embedding (B-H t-SNE) representations of kmer nucleotide frequencies (kmer size defined by <code>kmer_size</code>) using <code><a href='https://rdrr.io/pkg/Rtsne/man/Rtsne.html'>Rtsne</a></code> and merge the result. Additional arguments may be required for success (passed on through <code>...</code>), refer to the documentation of <code><a href='https://rdrr.io/pkg/Rtsne/man/Rtsne.html'>Rtsne</a></code>. This is done in parallel, thus setting the <code>num_threads</code> to the number of available cores may greatly increase the calculation time of large data. (<em>Default: </em> <code>FALSE</code>)</p></td>
    </tr>
    <tr>
      <th>kmer_size</th>
      <td><p>The kmer frequency size (k) used when <code>kmer_pca = TRUE</code> or <code>kmer_BH_tSNE = TRUE</code>. The default is tetramers (<code>k = 4</code>). (<em>Default: </em> <code>4</code>)</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>(<em>Logical</em>) Whether to print status messages during the loading process. (<em>Default: </em> <code>TRUE</code>)</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments are passed on to <code><a href='https://rdrr.io/pkg/Rtsne/man/Rtsne.html'>Rtsne</a></code>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A dataframe (tibble) compatible with other mmgenome2 functions.</p>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Kasper Skytte Andersen <a href='mailto:ksa@bio.aau.dk'>ksa@bio.aau.dk</a></p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> <span class='op'>(</span><span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>{</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://kasperskytte.github.io/mmgenome2'>mmgenome2</a></span><span class='op'>)</span>
<span class='va'>mm</span> <span class='op'>&lt;-</span> <span class='fu'>mmload</span><span class='op'>(</span>
  assembly <span class='op'>=</span> <span class='st'>"path/to/assembly.fa"</span>,
  coverage <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>
    nameofcoverage1 <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/utils/read.table.html'>read.csv</a></span><span class='op'>(</span><span class='st'>"path/to/coveragetable1.csv"</span>, col.names <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>,
    nameofcoverage2 <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/utils/read.table.html'>read.csv</a></span><span class='op'>(</span><span class='st'>"path/to/coveragetable2.csv"</span>, col.names <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
  <span class='op'>)</span>,
  essential_genes <span class='op'>=</span> <span class='st'>"path/to/ess_genes.txt"</span>,
  verbose <span class='op'>=</span> <span class='cn'>TRUE</span>
<span class='op'>)</span>
<span class='va'>mm</span>
<span class='op'>}</span>

</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Mads Albertsen, Kasper Skytte Andersen, Rasmus Hansen Kirkegaard, Søren Michael Karst.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script>
<script>
  docsearch({
    
    
    apiKey: '71d76a1c4adc57e7823b1c5a1e39065e',
    indexName: 'mmgenome2',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>



  </body>
</html>


