<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Visualise metagenomes in various ways — mmplot • mmgenome2</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>


<!-- docsearch -->
<script src="../docsearch.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous" />
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>



<meta property="og:title" content="Visualise metagenomes in various ways — mmplot" />
<meta property="og:description" content="Plots any information about the scaffolds contained in the given mm object as a scatterplot, for example different coverage variables, scaffold length, GC content, or anything else that were loaded with mmload. Scaffolds can then be highlighted and extracted using the locator and selection features." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-93913019-3"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93913019-3');
</script>


  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mmgenome2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/mmgenome2.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/kasperskytte/mmgenome2">
    <span class="fas fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
      <form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
        </div>
      </form>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Visualise metagenomes in various ways</h1>
    <small class="dont-index">Source: <a href='https://github.com/kasperskytte/mmgenome2/blob/master/R/mmplot.R'><code>R/mmplot.R</code></a></small>
    <div class="hidden name"><code>mmplot.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Plots any information about the scaffolds contained in the given mm object as a scatterplot, for example different coverage variables, scaffold length, GC content, or anything else that were loaded with <code><a href='mmload.html'>mmload</a></code>. Scaffolds can then be highlighted and extracted using the locator and selection features.</p>
    </div>

    <pre class="usage"><span class='fu'>mmplot</span><span class='op'>(</span>
  <span class='va'>mm</span>,
  <span class='va'>x</span>,
  <span class='va'>y</span>,
  min_length <span class='op'>=</span> <span class='fl'>0</span>,
  color_by <span class='op'>=</span> <span class='cn'>NULL</span>,
  locator <span class='op'>=</span> <span class='cn'>FALSE</span>,
  selection <span class='op'>=</span> <span class='cn'>NULL</span>,
  network <span class='op'>=</span> <span class='cn'>NULL</span>,
  shared_genes <span class='op'>=</span> <span class='cn'>FALSE</span>,
  label_scaffolds <span class='op'>=</span> <span class='cn'>FALSE</span>,
  label_scaffolds_by <span class='op'>=</span> <span class='st'>"scaffold"</span>,
  label_bins <span class='op'>=</span> <span class='cn'>NULL</span>,
  highlight_scaffolds <span class='op'>=</span> <span class='cn'>NULL</span>,
  highlight_color <span class='op'>=</span> <span class='st'>"darkred"</span>,
  x_scale <span class='op'>=</span> <span class='cn'>NULL</span>,
  x_limits <span class='op'>=</span> <span class='cn'>NULL</span>,
  y_scale <span class='op'>=</span> <span class='cn'>NULL</span>,
  y_limits <span class='op'>=</span> <span class='cn'>NULL</span>,
  alpha <span class='op'>=</span> <span class='fl'>0.1</span>,
  fixed_size <span class='op'>=</span> <span class='cn'>NULL</span>,
  size_scale <span class='op'>=</span> <span class='fl'>1</span>,
  factor_shape <span class='op'>=</span> <span class='st'>"outline"</span>,
  color_vector <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"blue"</span>, <span class='st'>"green"</span>, <span class='st'>"red"</span><span class='op'>)</span>,
  color_scale_log10 <span class='op'>=</span> <span class='cn'>FALSE</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>mm</th>
      <td><p>(<em>required</em>) A dataframe loaded with <code><a href='mmload.html'>mmload</a></code>.</p></td>
    </tr>
    <tr>
      <th>x</th>
      <td><p>(<em>required</em>) The variable from <code>mm</code> to plot on the first axis.</p></td>
    </tr>
    <tr>
      <th>y</th>
      <td><p>(<em>required</em>) The variable from <code>mm</code> to plot on the second axis.</p></td>
    </tr>
    <tr>
      <th>min_length</th>
      <td><p>Remove scaffolds with a length at or below this value before plotting. (<em>Default: </em> <code>0</code>)</p></td>
    </tr>
    <tr>
      <th>color_by</th>
      <td><p>Color the scaffolds by a variable in <code>mm</code>. (<em>Default: </em> <code>NULL</code>)</p></td>
    </tr>
    <tr>
      <th>locator</th>
      <td><p>(<em>Logical</em>) When <code>TRUE</code>, left-clicks in the plot are captured and the exact x/y-coordinates of the mouse clicks are returned. These coordinates can be used to highlight a selection of scaffolds in the plot, and also be used with <code><a href='mmextract.html'>mmextract</a></code> to extract all scaffolds within the selection from the data. (<em>Default: </em> <code>FALSE</code>)</p></td>
    </tr>
    <tr>
      <th>selection</th>
      <td><p>A 2-column dataframe with the x and y coordinates of points with which to draw a polygon onto the plot to highlight a selected region. A selection can be obtained by using the locator feature (by <code>locator = TRUE</code>). (<em>Default: </em> <code>NULL</code>)</p></td>
    </tr>
    <tr>
      <th>network</th>
      <td><p>Paired-end or mate-pair connections between scaffolds in long format. The first and second columns must contain all connected scaffold pairs and the third column the number of connections.</p></td>
    </tr>
    <tr>
      <th>shared_genes</th>
      <td><p>(<em>Logical</em>) If <code>TRUE</code>, lines will be drawn between scaffolds with any shared gene(s). (<em>Default: </em> <code>FALSE</code>)</p></td>
    </tr>
    <tr>
      <th>label_scaffolds</th>
      <td><p>Add text labels (with text from the variable in mm defined by <code>label_scaffolds_by</code>) to a selection of scaffolds by providing either a character vector of scaffold names, or a dataframe with scaffold names in the first column. If set to <code>TRUE</code> then <em>all</em> scaffolds will be labelled. (<em>Default: </em> <code>FALSE</code>)</p></td>
    </tr>
    <tr>
      <th>label_scaffolds_by</th>
      <td><p>The variable in mm by which to label the scaffolds defined by <code>label_scaffolds</code>. (<em>Default: </em> <code>"scaffold"</code>)</p></td>
    </tr>
    <tr>
      <th>label_bins</th>
      <td><p>Add labels at the centroids of bins (groups of scaffolds) defined by a variable in <code>mm</code>. (<em>Default: </em> <code>NULL</code>)</p></td>
    </tr>
    <tr>
      <th>highlight_scaffolds</th>
      <td><p>A vector of scaffold names or a dataframe loaded with <code><a href='mmload.html'>mmload</a></code> containing scaffolds to highlight in the plot with the color set by <code>highlight_color</code>. (<em>Default: </em> <code>NULL</code>)</p></td>
    </tr>
    <tr>
      <th>highlight_color</th>
      <td><p>The color with which to highlight the scaffolds set by <code>highlight</code>. (<em>Default: </em> <code>"darkred"</code>)</p></td>
    </tr>
    <tr>
      <th>x_scale</th>
      <td><p>Log10-scale (<code>"log10"</code>) or square-root scale <code>"sqrt"</code> the x axis. (<em>Default: </em> <code>NULL</code>)</p></td>
    </tr>
    <tr>
      <th>x_limits</th>
      <td><p>Axis limits of the x axis. Must be a vector of length 2 where the first number is the lower limit and the second number is the upper limit. Use <code>NA</code> to automatically detect the lowest and highest values, respectively. (<em>Default: </em> <code>NULL</code>)</p></td>
    </tr>
    <tr>
      <th>y_scale</th>
      <td><p>Log10-scale (<code>"log10"</code>) or square-root scale <code>"sqrt"</code> the y axis. (<em>Default: </em> <code>NULL</code>)</p></td>
    </tr>
    <tr>
      <th>y_limits</th>
      <td><p>Axis limits of the y axis. Must be a vector of length 2 where the first number is the lower limit and the second number is the upper limit. Use <code>NA</code> to automatically detect the lowest and highest values, respectively. (<em>Default: </em> <code>NULL</code>)</p></td>
    </tr>
    <tr>
      <th>alpha</th>
      <td><p>The transparancy of the scaffold points, where 0 is invisible and 1 is opaque. (<em>Default: </em> <code>0.1</code>)</p></td>
    </tr>
    <tr>
      <th>fixed_size</th>
      <td><p>A fixed size for all scaffolds if set. If <code>NULL</code> then the scaffolds are scaled by length. (<em>Default: </em> <code>NULL</code>)</p></td>
    </tr>
    <tr>
      <th>size_scale</th>
      <td><p>A factor to scale the sizes of the scaffolds plotted. Only applies when <code>fixed_size</code> is set to <code>NULL</code> and the scaffolds are scaled by length. (<em>Default: </em> <code>1</code>)</p></td>
    </tr>
    <tr>
      <th>factor_shape</th>
      <td><p>When <code>color_by</code> is a categorical variable (factor or character) then set the shape of the scaffolds to either <code>"solid"</code> or <code>"outline"</code>. (<em>Default: </em> <code>"outline"</code>)</p></td>
    </tr>
    <tr>
      <th>color_vector</th>
      <td><p>The colors from which to generate a color gradient when <code>color_by</code> is set and the variable is continuous. Any number of colors can be used. (<em>Default: </em> <code><a href='https://rdrr.io/r/base/c.html'>c("blue", "green", "red")</a></code>)</p></td>
    </tr>
    <tr>
      <th>color_scale_log10</th>
      <td><p>(<em>Logical</em>) Log10-scale the color gradient when <code>color_by</code> is set and the variable is continuous. (<em>Default: </em> <code>FALSE</code>)</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A ggplot object. Note that mmgenome2 hides all warnings produced by ggplot objects.</p>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Kasper Skytte Andersen <a href='mailto:ksa@bio.aau.dk'>ksa@bio.aau.dk</a></p>
<p>Soren M. Karst <a href='mailto:smk@bio.aau.dk'>smk@bio.aau.dk</a></p>
<p>Mads Albertsen <a href='mailto:MadsAlbertsen85@gmail.com'>MadsAlbertsen85@gmail.com</a></p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://kasperskytte.github.io/mmgenome2'>mmgenome2</a></span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>mmgenome2</span><span class='op'>)</span>
<span class='va'>mmgenome2</span>
</div><div class='output co'>#&gt; <span style='color: #949494;'># A tibble: 97,285 x 13</span><span>
#&gt;    scaffold length    gc cov_C13.11.14 cov_C13.11.25 cov_C13.12.03 cov_C14.01.09
#&gt;    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>         </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>         </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>         </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>         </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>
#&gt; </span><span style='color: #BCBCBC;'> 1</span><span> 1          </span><span style='text-decoration: underline;'>8</span><span>264  57.8       1.44             53.6         0             0.066
#&gt; </span><span style='color: #BCBCBC;'> 2</span><span> 2          </span><span style='text-decoration: underline;'>1</span><span>027  57.0       0.625            24.2         0             0    
#&gt; </span><span style='color: #BCBCBC;'> 3</span><span> 3          </span><span style='text-decoration: underline;'>1</span><span>665  55.9      13.5             434.          0.166         0.177
#&gt; </span><span style='color: #BCBCBC;'> 4</span><span> 4          </span><span style='text-decoration: underline;'>9</span><span>056  35.9       0.01             23.4         0             0    
#&gt; </span><span style='color: #BCBCBC;'> 5</span><span> 5          </span><span style='text-decoration: underline;'>3</span><span>343  64.0       3.20             16.4         0             0    
#&gt; </span><span style='color: #BCBCBC;'> 6</span><span> 6         </span><span style='text-decoration: underline;'>98</span><span>207  39.1       0.009</span><span style='text-decoration: underline;'>66</span><span>          24.5         3.29          9.85 
#&gt; </span><span style='color: #BCBCBC;'> 7</span><span> 7          </span><span style='text-decoration: underline;'>6</span><span>480  63.0       2.61             19.2         1.46         12.3  
#&gt; </span><span style='color: #BCBCBC;'> 8</span><span> 8         </span><span style='text-decoration: underline;'>15</span><span>790  61.7       2.78             21.2         1.62         10.3  
#&gt; </span><span style='color: #BCBCBC;'> 9</span><span> 9          </span><span style='text-decoration: underline;'>1</span><span>403  70.4      85.1             192.          0             0    
#&gt; </span><span style='color: #BCBCBC;'>10</span><span> 10         </span><span style='text-decoration: underline;'>2</span><span>018  70.2      50.3             101.          0             0    
#&gt; </span><span style='color: #949494;'># … with 97,275 more rows, and 6 more variables: PC1 &lt;dbl&gt;, PC2 &lt;dbl&gt;,</span><span>
#&gt; </span><span style='color: #949494;'>#   PC3 &lt;dbl&gt;, geneID &lt;chr&gt;, taxonomy &lt;fct&gt;, rRNA16S &lt;fct&gt;</span><span></div><div class='input'><span class='fu'>mmplot</span><span class='op'>(</span><span class='va'>mmgenome2</span>,
  min_length <span class='op'>=</span> <span class='fl'>3000</span>,
  x <span class='op'>=</span> <span class='st'>"cov_C13.11.25"</span>,
  y <span class='op'>=</span> <span class='st'>"cov_C14.01.09"</span>,
  color_by <span class='op'>=</span> <span class='st'>"taxonomy"</span>,
  <span class='co'># locator = TRUE,</span>
  x_scale <span class='op'>=</span> <span class='st'>"log10"</span>,
  y_scale <span class='op'>=</span> <span class='st'>"log10"</span>
<span class='op'>)</span>
</div><div class='img'><img src='mmplot-1.png' alt='' width='700' height='433' /></div><div class='input'><span class='co'># Set "locator = TRUE" to interactively capture the coordinates of</span>
<span class='co'># mouse clicks in an mmplot, or provide coordinates with "selection":</span>
<span class='va'>selection</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>
  cov_C13.11.25 <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>7.2</span>, <span class='fl'>16.2</span>, <span class='fl'>25.2</span>, <span class='fl'>23.3</span>, <span class='fl'>10.1</span><span class='op'>)</span>,
  cov_C14.01.09 <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>47</span>, <span class='fl'>77</span>, <span class='fl'>52.8</span>, <span class='fl'>29.5</span>, <span class='fl'>22.1</span><span class='op'>)</span>
<span class='op'>)</span>
<span class='fu'>mmplot</span><span class='op'>(</span><span class='va'>mmgenome2</span>,
  min_length <span class='op'>=</span> <span class='fl'>10000</span>,
  x <span class='op'>=</span> <span class='st'>"cov_C13.11.25"</span>,
  y <span class='op'>=</span> <span class='st'>"cov_C14.01.09"</span>,
  color_by <span class='op'>=</span> <span class='st'>"taxonomy"</span>,
  x_scale <span class='op'>=</span> <span class='st'>"log10"</span>,
  y_scale <span class='op'>=</span> <span class='st'>"log10"</span>,
  x_limits <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='cn'>NA</span><span class='op'>)</span>, <span class='co'># zoom in at minimum 1x coverage</span>
  y_limits <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='cn'>NA</span><span class='op'>)</span>, <span class='co'># zoom in at minimum 1x coverage</span>
  selection <span class='op'>=</span> <span class='va'>selection</span>
<span class='op'>)</span> <span class='co'># highlight the selection marked with locator</span>
</div><div class='img'><img src='mmplot-2.png' alt='' width='700' height='433' /></div></span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Mads Albertsen, Kasper Skytte Andersen, Rasmus Hansen Kirkegaard, Søren Michael Karst.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script>
<script>
  docsearch({
    
    
    apiKey: '71d76a1c4adc57e7823b1c5a1e39065e',
    indexName: 'mmgenome2',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>



  </body>
</html>


