<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Network plot of mate-pair or paired-end connected scaffold reads — mmnetwork • mmgenome2</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.1/css/all.min.css" integrity="sha256-PbSmjxuVAzJ6FPvNYsrXygfGhNJYyZ2GktDbkMBqQZg=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.1/css/v4-shims.min.css" integrity="sha256-A6jcAdwFD48VMjlI3GDxUd+eCQa7/KWy6G9oe/ovaPA=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>


<!-- docsearch -->
<script src="../docsearch.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous" />
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>



<meta property="og:title" content="Network plot of mate-pair or paired-end connected scaffold reads — mmnetwork" />
<meta property="og:description" content="Plots all connected scaffolds in a network plot. Scaffolds can then be highlighted and extracted using the locator and selection features." />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-93913019-3"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93913019-3');
</script>


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mmgenome2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.0.14</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/mmgenome2.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/kasperskytte/mmgenome2">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
      <form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
        </div>
      </form>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Network plot of mate-pair or paired-end connected scaffold reads</h1>
    
    <div class="hidden name"><code>mmnetwork.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Plots all connected scaffolds in a network plot. Scaffolds can then be highlighted and extracted using the locator and selection features.</p>
    </div>

    <pre class="usage"><span class='fu'>mmnetwork</span>(<span class='no'>mm</span>, <span class='no'>network</span>, <span class='kw'>min_connections</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>color_by</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>color_scale_log10</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>locator</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>selection</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>highlight_labels</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>highlight_color</span> <span class='kw'>=</span> <span class='st'>"darkred"</span>,
  <span class='kw'>links_scale</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>scaffold_labels</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>print_nolinks</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>seed</span> <span class='kw'>=</span> <span class='fl'>0</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>mm</th>
      <td><p>(<em>required</em>) A dataframe loaded with <code><a href='mmload.html'>mmload</a></code>.</p></td>
    </tr>
    <tr>
      <th>network</th>
      <td><p>(<em>required</em>) Paired-end or mate-pair connections between scaffolds in long format. The first and second columns must contain all connected scaffold pairs and the third column the number of connections.</p></td>
    </tr>
    <tr>
      <th>min_connections</th>
      <td><p>Filter all scaffold pairs with equal to or less than this number of connections. (<em>Default: </em> <code>2</code>)</p></td>
    </tr>
    <tr>
      <th>color_by</th>
      <td><p>Color the scaffolds by a variable in <code>mm</code>. (<em>Default: </em> <code>NULL</code>)</p></td>
    </tr>
    <tr>
      <th>color_scale_log10</th>
      <td><p>(<em>Logical</em>) Log10-scale the color gradient when <code>color_by</code> is set and the variable is continuous (<em>Default: </em> <code>FALSE</code>)</p></td>
    </tr>
    <tr>
      <th>locator</th>
      <td><p>(<em>Logical</em>) When <code>TRUE</code>, left-clicks in the plot are captured and the exact x/y-coordinates of the mouse clicks are captured and returned. These coordinates can be used to highlight a selection of scaffolds in the plot, and to extract all scaffolds within the selection. (<em>Default: </em> <code>FALSE</code>)</p></td>
    </tr>
    <tr>
      <th>selection</th>
      <td><p>A 2-column dataframe with the x and y coordinates of points with which to draw a polygon onto the plot to highlight a selected region. A selection can be obtained by using the locator feature (by <code>locator = TRUE</code>). (<em>Default: </em> <code>NULL</code>)</p></td>
    </tr>
    <tr>
      <th>highlight_labels</th>
      <td><p>A dataframe or vector of scaffold names whose labels to highlight in the plot (colored by <code>highlight_color</code>). (<em>Default: </em> <code>NULL</code>)</p></td>
    </tr>
    <tr>
      <th>highlight_color</th>
      <td><p>The color with which to highlight the scaffold labels defined by <code>highlight_labels</code>. (<em>Default: </em> <code>"darkred"</code>)</p></td>
    </tr>
    <tr>
      <th>links_scale</th>
      <td><p>A factor to scale the sizes of the links plotted between scaffolds. (<em>Default: </em> <code>1</code>)</p></td>
    </tr>
    <tr>
      <th>scaffold_labels</th>
      <td><p>Add labels with the scaffold names of <em>all</em> scaffolds.</p></td>
    </tr>
    <tr>
      <th>print_nolinks</th>
      <td><p>(<em>Logical</em>) Print the names of all scaffolds with no links to other scaffolds to the console. (<em>Default: </em> <code>FALSE</code>)</p></td>
    </tr>
    <tr>
      <th>seed</th>
      <td><p>Network plots are based on Random Number Generation, and this is used to set a specific random seed (with <code>set.seed</code>) allowing for reproducible network plots. Set to NULL to generate a unique network plot with each run of <code>mmnetwork</code>. (<em>Default: </em> <code>0</code>)</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A ggplot object. Note that mmgenome2 hides all warnings produced by ggplot objects.</p>
    <h2 class="hasAnchor" id="using-the-locator-with-mmnetwork"><a class="anchor" href="#using-the-locator-with-mmnetwork"></a>Using the locator with mmnetwork</h2>

    

<p>Unlike a plot generated by <code><a href='mmplot.html'>mmplot</a></code>, the coordinates of points in each dimension in a <code>mmnetwork</code> plot is not contained within the provided dataframe <code>mm</code> itself. As such <code>mmnetwork</code> does not have a corresponding extract function like <code><a href='mmextract.html'>mmextract</a></code> is for <code><a href='mmplot.html'>mmplot</a></code>. Instead, a subset of <code>mm</code> containing all scaffolds within the selection polygon is available in the returned ggplot object with <code>plot$data_in_selection</code>.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/library.html'>library</a></span>(<span class='no'>mmgenome2</span>)
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='no'>mmgenome2</span>)
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='no'>paired_ends</span>)
<span class='no'>mmgenome2</span></div><div class='output co'>#&gt; <span style='color: #949494;'># A tibble: 97,285 x 13</span><span>
#&gt;    scaffold length    gc cov_C13.11.14 cov_C13.11.25 cov_C13.12.03 cov_C14.01.09
#&gt;    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>         </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>         </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>         </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>         </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>
#&gt; </span><span style='color: #BCBCBC;'> 1</span><span> 1          </span><span style='text-decoration: underline;'>8</span><span>264  57.8       1.44             53.6         0             0.066
#&gt; </span><span style='color: #BCBCBC;'> 2</span><span> 2          </span><span style='text-decoration: underline;'>1</span><span>027  57.0       0.625            24.2         0             0    
#&gt; </span><span style='color: #BCBCBC;'> 3</span><span> 3          </span><span style='text-decoration: underline;'>1</span><span>665  55.9      13.5             434.          0.166         0.177
#&gt; </span><span style='color: #BCBCBC;'> 4</span><span> 4          </span><span style='text-decoration: underline;'>9</span><span>056  35.9       0.01             23.4         0             0    
#&gt; </span><span style='color: #BCBCBC;'> 5</span><span> 5          </span><span style='text-decoration: underline;'>3</span><span>343  64.0       3.20             16.4         0             0    
#&gt; </span><span style='color: #BCBCBC;'> 6</span><span> 6         </span><span style='text-decoration: underline;'>98</span><span>207  39.1       0.009</span><span style='text-decoration: underline;'>66</span><span>          24.5         3.29          9.85 
#&gt; </span><span style='color: #BCBCBC;'> 7</span><span> 7          </span><span style='text-decoration: underline;'>6</span><span>480  63.0       2.61             19.2         1.46         12.3  
#&gt; </span><span style='color: #BCBCBC;'> 8</span><span> 8         </span><span style='text-decoration: underline;'>15</span><span>790  61.7       2.78             21.2         1.62         10.3  
#&gt; </span><span style='color: #BCBCBC;'> 9</span><span> 9          </span><span style='text-decoration: underline;'>1</span><span>403  70.4      85.1             192.          0             0    
#&gt; </span><span style='color: #BCBCBC;'>10</span><span> 10         </span><span style='text-decoration: underline;'>2</span><span>018  70.2      50.3             101.          0             0    
#&gt; </span><span style='color: #949494;'># … with 97,275 more rows, and 6 more variables: PC1 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, PC2 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>,
#&gt; #   PC3 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, geneID </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span style='color: #949494;'>, taxonomy </span><span style='color: #949494;font-style: italic;'>&lt;fct&gt;</span><span style='color: #949494;'>, rRNA16S </span><span style='color: #949494;font-style: italic;'>&lt;fct&gt;</span><span></div><div class='input'><span class='no'>selection</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span>(
  <span class='kw'>cov_C13.11.25</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fl'>7.2</span>, <span class='fl'>16.2</span>, <span class='fl'>25.2</span>, <span class='fl'>23.3</span>, <span class='fl'>10.1</span>),
  <span class='kw'>cov_C14.01.09</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fl'>47</span>, <span class='fl'>77</span>, <span class='fl'>52.8</span>, <span class='fl'>29.5</span>, <span class='fl'>22.1</span>)
)
<span class='no'>mmgenome2_extraction</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='mmextract.html'>mmextract</a></span>(<span class='no'>mmgenome2</span>, <span class='kw'>selection</span> <span class='kw'>=</span> <span class='no'>selection</span>)</div><div class='output co'>#&gt; <span class='message'>61 scaffolds (or 1.34% of the scaffolds in mm, weighted by length) remain after 97224 of 97285 scaffolds have been filtered.</span></div><div class='input'><span class='no'>mmgenome2_extraction</span></div><div class='output co'>#&gt; </span><span style='color: #949494;'># A tibble: 61 x 13</span><span>
#&gt;    scaffold length    gc cov_C13.11.14 cov_C13.11.25 cov_C13.12.03 cov_C14.01.09
#&gt;    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>         </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>         </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>         </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>         </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>
#&gt; </span><span style='color: #BCBCBC;'> 1</span><span> 180       </span><span style='text-decoration: underline;'>22</span><span>093  74.0         0.298          12.0          7.34          37.9
#&gt; </span><span style='color: #BCBCBC;'> 2</span><span> 181        </span><span style='text-decoration: underline;'>2</span><span>970  68.0         0.34           13.9          7.37          38.6
#&gt; </span><span style='color: #BCBCBC;'> 3</span><span> 227      </span><span style='text-decoration: underline;'>190</span><span>826  65.4         0.348          13.0          6.91          38.5
#&gt; </span><span style='color: #BCBCBC;'> 4</span><span> 253      </span><span style='text-decoration: underline;'>296</span><span>715  70.2         1.06           15.2          7.00          41.1
#&gt; </span><span style='color: #BCBCBC;'> 5</span><span> 315      </span><span style='text-decoration: underline;'>218</span><span>862  71.7         0.366          12.6          7.17          38.8
#&gt; </span><span style='color: #BCBCBC;'> 6</span><span> 343       </span><span style='text-decoration: underline;'>27</span><span>203  57.6         0.235          10.7          5.07          32.0
#&gt; </span><span style='color: #BCBCBC;'> 7</span><span> 381        </span><span style='text-decoration: underline;'>6</span><span>128  67.2         8.11           22.6          5.89          48.7
#&gt; </span><span style='color: #BCBCBC;'> 8</span><span> 386      </span><span style='text-decoration: underline;'>164</span><span>942  68.2         0.359          12.3          6.73          37.2
#&gt; </span><span style='color: #BCBCBC;'> 9</span><span> 396        </span><span style='text-decoration: underline;'>7</span><span>448  72.6         0.329          13.0          7.59          41.0
#&gt; </span><span style='color: #BCBCBC;'>10</span><span> 423      </span><span style='text-decoration: underline;'>100</span><span>199  68.7         1.40           15.3          6.71          40.0
#&gt; </span><span style='color: #949494;'># … with 51 more rows, and 6 more variables: PC1 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, PC2 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, PC3 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>,
#&gt; #   geneID </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span style='color: #949494;'>, taxonomy </span><span style='color: #949494;font-style: italic;'>&lt;fct&gt;</span><span style='color: #949494;'>, rRNA16S </span><span style='color: #949494;font-style: italic;'>&lt;fct&gt;</span><span></div><div class='input'><span class='no'>p</span> <span class='kw'>&lt;-</span> <span class='fu'>mmnetwork</span>(<span class='no'>mmgenome2_extraction</span>,
  <span class='kw'>network</span> <span class='kw'>=</span> <span class='no'>paired_ends</span>,
  <span class='kw'>min_connections</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>color_by</span> <span class='kw'>=</span> <span class='st'>"taxonomy"</span>
)
<span class='no'>p</span></div><div class='img'><img src='mmnetwork-1.png' alt='' width='700' height='433' /></div></span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#using-the-locator-with-mmnetwork">Using the locator with mmnetwork</a></li>
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Kasper Skytte Andersen <a href='mailto:ksa@bio.aau.dk'>ksa@bio.aau.dk</a></p>
<p>Soren M. Karst <a href='mailto:smk@bio.aau.dk'>smk@bio.aau.dk</a></p>
<p>Mads Albertsen <a href='mailto:MadsAlbertsen85@gmail.com'>MadsAlbertsen85@gmail.com</a></p>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Mads Albertsen, Kasper Skytte Andersen, Rasmus Hansen Kirkegaard, Søren Michael Karst.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.9000.</p>
</div>

      </footer>
   </div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script>
<script>
  docsearch({
    
    
    apiKey: '71d76a1c4adc57e7823b1c5a1e39065e',
    indexName: 'mmgenome2',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>



  </body>
</html>


