<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Print basic statistics of metagenome data — mmstats • mmgenome2</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>


<!-- docsearch -->
<script src="../docsearch.js"></script>
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/docsearch.js@2/dist/cdn/docsearch.min.css" />
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>


<meta property="og:title" content="Print basic statistics of metagenome data — mmstats" />

<meta property="og:description" content="Calculates some basic statistics like the N50 length, minimum-, mean-, and maximum scaffold lengths, mean GC content, information about essential genes, and more..." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-93913019-3"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93913019-3');
</script>

  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mmgenome2</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.0.12</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/mmgenome2.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/kasperskytte/mmgenome2">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
      <form class="navbar-form navbar-right" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
        </div>
      </form>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Print basic statistics of metagenome data</h1>
    
    <div class="hidden name"><code>mmstats.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Calculates some basic statistics like the N50 length, minimum-, mean-, and maximum scaffold lengths, mean GC content, information about essential genes, and more...</p>
    
    </div>

    <pre class="usage"><span class='fu'>mmstats</span>(<span class='no'>mm</span>, <span class='kw'>original_data</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>print</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>mm</th>
      <td><p>(<em>required</em>) A dataframe loaded with <code><a href='mmload.html'>mmload</a></code>.</p></td>
    </tr>
    <tr>
      <th>original_data</th>
      <td><p>If <code>mm</code> is a subset/extraction of another dataframe loaded with <code><a href='mmload.html'>mmload</a></code>, then provide here the original dataframe from which the extraction originates to compare the extraction to the original data, see examples. (<em>Default: </em> <code>NULL</code>)</p></td>
    </tr>
    <tr>
      <th>print</th>
      <td><p>(<em>logical</em>) Whether to print the calculated stats or not. (<em>Default: </em> <code>TRUE</code>)</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A dataframe with the calculated stats is returned invisibly.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The returned stats are calculated as follows:</p><ul>
<li><p>Scaffolds: The number of different scaffolds in the assembly.</p></li>
<li><p>N50: The shortest sequence length at 50<!-- % of the assembly. --></p></li>
<li><p>Length.total: The total size of the assembly.</p></li>
<li><p>Length.max: The size of the largest scaffold in the assembly.</p></li>
<li><p>Length.mean: The average scaffold size in the assembly.</p></li>
<li><p>Length.min: The size of the smallest scaffold in the assembly.</p></li>
<li><p>weighted_GC_mean: The average GC content in the assembly, weighted by scaffold sizes.</p></li>
<li><p>cov_*: The average coverage of each coverage variable in mm, weighted by scaffold sizes. (Only columns starting with "cov_" will be shown)</p></li>
<li><p>Ess.genes.total: The total number of essential genes, if any have been loaded.</p></li>
<li><p>Ess.genes.unique: The number of unique essential genes, if any have been loaded.</p></li>
</ul>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>library</span>(<span class='no'>mmgenome2</span>)
<span class='fu'>data</span>(<span class='no'>mmgenome2</span>)
<span class='fu'>mmstats</span>(<span class='no'>mmgenome2</span>)</div><div class='output co'>#&gt;                   General stats
#&gt; Scaffolds             97 285   
#&gt; N50                    5 291   
#&gt; Length.total     331 908 376   
#&gt; Length.max         1 446 979   
#&gt; Length.mean            3 411.71
#&gt; Length.min             1 000   
#&gt; weighted_GC_mean          52.22
#&gt; cov_C13.11.14              2.62
#&gt; cov_C13.11.25             16.06
#&gt; cov_C13.12.03              2.91
#&gt; cov_C14.01.09             11.58
#&gt; Ess.genes.total        7 138   
#&gt; Ess.genes.unique         109   </div><div class='input'>
<span class='co'># Compare an extraction with the original data from which the extraction originates:</span>
<span class='no'>selection</span> <span class='kw'>&lt;-</span> <span class='fu'>data.frame</span>(
  <span class='kw'>cov_C13.11.25</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='fl'>7.2</span>, <span class='fl'>16.2</span>, <span class='fl'>25.2</span>, <span class='fl'>23.3</span>, <span class='fl'>10.1</span>),
  <span class='kw'>cov_C14.01.09</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='fl'>47</span>, <span class='fl'>77</span>, <span class='fl'>52.8</span>, <span class='fl'>29.5</span>, <span class='fl'>22.1</span>)
)
<span class='no'>mmgenome2_extraction</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='mmextract.html'>mmextract</a></span>(<span class='no'>mmgenome2</span>,
  <span class='kw'>selection</span> <span class='kw'>=</span> <span class='no'>selection</span>
)</div><div class='output co'>#&gt; <span class='message'>61 scaffolds (or 1.34% of the scaffolds in mm, weighted by length) remain after 97224 of 97285 scaffolds have been filtered.</span></div><div class='input'>
<span class='fu'>mmstats</span>(<span class='no'>mmgenome2_extraction</span>, <span class='kw'>original_data</span> <span class='kw'>=</span> <span class='no'>mmgenome2</span>)</div><div class='output co'>#&gt;                               General stats
#&gt; Scaffolds                             61   
#&gt; N50                              199 460   
#&gt; Length.total                   4 452 901   
#&gt; Length.max                       454 223   
#&gt; Length.mean                       72 998.38
#&gt; Length.min                         1 043   
#&gt; weighted_GC_mean                      70.96
#&gt; cov_C13.11.14 (% of original)          0.3 
#&gt; cov_C13.11.25 (% of original)          1.11
#&gt; cov_C13.12.03 (% of original)          3.19
#&gt; cov_C14.01.09 (% of original)          4.49
#&gt; Ess.genes.total                      109   
#&gt; Ess.genes.unique                     105   </div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Kasper Skytte Andersen <a href='mailto:ksa@bio.aau.dk'>ksa@bio.aau.dk</a>

Soren M. Karst <a href='mailto:smk@bio.aau.dk'>smk@bio.aau.dk</a>

Mads Albertsen <a href='mailto:MadsAlbertsen85@gmail.com'>MadsAlbertsen85@gmail.com</a>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Mads Albertsen, Kasper Skytte Andersen, Rasmus Hansen Kirkegaard, Søren Michael Karst.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  
<script type="text/javascript" src="https://cdn.jsdelivr.net/npm/docsearch.js@2/dist/cdn/docsearch.min.js"></script>
<script>
  docsearch({
    
    
    apiKey: '71d76a1c4adc57e7823b1c5a1e39065e',
    indexName: 'mmgenome2',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>


  </body>
</html>

